doc/Subprocess debugging.md

```diff
@@ -0,0 +1,318 @@
+# Subprocess debugging
+
+## Terminology
+
+_Debuggee process_ - the process that is being debugged.
+
+_IDE_ - VSCode or other DAP client.
+
+_Debug server_ - pydevd with ptvsd wrapper; hosted inside the debuggee process,
+one for each.
+
+_Debug adapter_ - ptvsd adapter that mediates between IDE and server.
+
+_IDE listener port_ - port opened by the adapter, on which it listens for incoming
+connections from the IDE.
+
+_Server listener port_ - port opened by the adapter, on which it listens for incoming
+connections from the servers.
+
+_Adapter listener port_ - port opened by the server, on which it listens for incoming
+connection from the adapter.
+
+
+
+## "launch" scenario
+
+1. User starts debugging (F5) with "launch" debug config.
+1. User code spawns child process.
+1. User stops debugging.
+
+```mermaid
+sequenceDiagram
+# Install "GitHub + Mermaid" from the Chrome Web Store to render the diagram
+
+participant IDE
+participant Adapter
+participant Debuggee_1
+participant Debuggee_2
+
+
+Note left of IDE: user starts<br/>debugging
+
+IDE ->> Adapter: spawn and connect over stdio
+
+IDE ->>+ Adapter: request "launch"
+
+Adapter ->>+ Debuggee_1: spawn and pass server listener port (cmdline)
+
+Debuggee_1 -->>- Adapter: connect to server listener port
+
+Adapter ->>+ Debuggee_1: request "initialize", "launch"
+
+Debuggee_1 -->>- Adapter: respond to "initialize", "launch"
+
+Debuggee_1 ->>+ Debuggee_1: start debug session
+
+Adapter -->>- IDE: respond to "launch"
+
+loop every message between IDE and Debuggee_1
+Note over IDE,Debuggee_1: propagate message
+end
+
+
+Note right of Debuggee_1: user code spawns<br/>child process
+
+Debuggee_1 ->>+ Debuggee_2: spawn and pass server listener port (cmdline)
+
+Debuggee_2 ->>- Adapter: connect to server listener port
+
+Adapter ->>+ Debuggee_2: request "pydevd_systemInfo"
+
+Debuggee_2 -->>- Adapter: respond to "pydevd_systemInfo"
+
+Adapter ->>+ IDE: "ptvsd_subprocess" event:
+
+IDE ->>- Adapter: connect to IDE listener port
+
+IDE ->>+ Adapter: request "attach" to Debuggee_2
+
+Adapter ->>+ Debuggee_2: request "initialize", "attach"
+
+Debuggee_2 -->>- Adapter: respond to "initialize", "attach"
+
+Debuggee_2 ->>+ Debuggee_2: start debug session
+
+Adapter -->>- IDE: respond to "attach"
+
+loop every message between IDE and Debuggee_2
+Note over IDE,Debuggee_2: propagate message
+end
+
+
+Note left of IDE: user stops debugging
```

Comment:
When the user `stops debugging`, if it does a disconnect to the `Debugee_2`, it'll resume the program if it was paused in any breakpoint... is this intended? (if that's the case, the note should probably be `user stops debugging Debugee_2`)

Reply:
The scenario here is that user hits the Stop button in the IDE. VSCode has the notion of a tree of processes in the debug session, and so long as we report parent and children properly, doing that will always disconnect and terminate the entire tree.

There's a separate scenario where they open the process list, and disconnect specifically from that one process. It's not fully captured in this diagram, but the expected behavior is for it to continue running - same as it would if it were a single-process attach scenario.

Reply:
Ok, so, it seems that sometimes disconnect should run user code and sometimes it shouldn't.

I've created https://github.com/microsoft/ptvsd/issues/1837 to address this in `pydevd` (we should probably update the diagram to be explicit about that message after we decide how this should be handled so that user code is not run on disconnect).

Reply:
Resolving since it should be discussed in #1837.